src/main.cpp

```diff
@@ -2789,6 +2791,80 @@ bool TestBlockValidity(CValidationState &state, const CBlock& block, CBlockIndex
 
 
 
+void ClearBlockFileInfo(int nFile)
+{
+    LOCK(cs_main);
+    if (vinfoBlockFile[nFile].nUndoSize)
+        vinfoBlockFile[nFile].nSize = 0;
+    else
+        vinfoBlockFile[nFile].SetNull();
+    setDirtyFileInfo.insert(nFile);
+}
+
+void ClearUndoFileInfo(int nFile)
+{
+    LOCK(cs_main);
+    if (vinfoBlockFile[nFile].nSize)
+        vinfoBlockFile[nFile].nUndoSize = 0;
+    else
+        vinfoBlockFile[nFile].SetNull();
+    setDirtyFileInfo.insert(nFile);
+}
+
+enum BlockOrUndo {
+    BLOCK = false,
+    UNDO = true
+};
+
+bool RemoveDiskFile(int nFile, BlockOrUndo blockOrUndo)
+{
+    LOCK(cs_main);
+    CDiskBlockPos pos(nFile, 0);
+    const char* prefix = blockOrUndo ? "rev" : "blk";
+    if (boost::filesystem::remove(GetBlockPosFilename(pos, prefix))) {
+        LogPrintf("File %s removed\n", GetBlockPosFilename(pos, prefix));
+        blockOrUndo ? ClearUndoFileInfo(nFile) : ClearBlockFileInfo(nFile);
+        return true;
+    }
+    LogPrintf("Error removing file %s\n", GetBlockPosFilename(pos, prefix));
+    return false;
+}
+
+bool RemoveBlockFile(int nFile)
+{
+    return RemoveDiskFile(nFile, BLOCK);
+}
+
+bool RemoveUndoFile(int nFile)
+{
+    return RemoveDiskFile(nFile, UNDO);
+}
+
+
+
+bool PruneBlockFiles()
+{
+    if (!CheckBlockFiles())
+        return false;
+    if (setDataFilePrunable.empty() && setUndoFilePrunable.empty()) {
+        LogPrintf("There's nothing to prune.\n");
+        return false;
+    }
+    bool fFileRemoved = false;
+    int dataPrunable = *setDataFilePrunable.begin(), undoPrunable = *setUndoFilePrunable.begin();
```

Comment:
Is it guaranteed that both setDataFilePrunable and setUndoFilePrunable are both non-empty at this stage? The test above only guarantees that at least one of them is non-empty. Don't dereference begin() of an empty set.